igann/igann.py

```diff
@@ -586,33 +590,68 @@ def _print_results(self, counter, counter_no_progress, eval, boost_rate, train_l
                     new_best_symb, counter, boost_rate, train_loss, val_loss))
 
     def _optimize_classification_threshold(self, X_train, y_train):
-        '''
-        This function optimizes the classification threshold for the training set for later predictions.
-        The use of the function is triggered by setting the parameter optimize_threshold to True.
-        This is one method which does the job. However, we noticed that it is not always the best method and hence it
-        defaults to no threshold optimization.
-        '''
-
-        y_proba = self.predict_raw(X_train)
-
-        # detach and numpy
-        y_proba = y_proba.detach().cpu().numpy()
-        y_train = y_train.detach().cpu().numpy()
-        fpr, tpr, trs = roc_curve(y_train, y_proba)
-
-        roc_scores = []
-        thresholds = []
-        for thres in trs:
-            thresholds.append(thres)
-            y_pred = np.where(y_proba > thres, 1, -1)
-            # Apply desired utility function to y_preds, for example accuracy.
-            roc_scores.append(roc_auc_score(y_train.squeeze(), y_pred.squeeze()))
-        # convert roc_scores to numpy array
-        roc_scores = np.array(roc_scores)
-        # get the index of the best threshold
-        ix = np.argmax(roc_scores)
-        # get the best threshold
-        return thresholds[ix]
+        """
+        Optimize the classification threshold for the training set by selecting the threshold that minimizes
+        the Euclidean distance to the perfect classification point (FPR = 0, TPR = 1).
+        This aims for an optimal combination of low false positives and high true positives.
+        """
+        y_proba = self.predict_proba(X_train)[:, 1]
+
+        # Convert tensors to numpy arrays if necessary (applicable if using PyTorch or TensorFlow)
+        if hasattr(y_proba, "detach"):
+            y_proba = y_proba.detach().cpu().numpy()
+        if hasattr(y_train, "detach"):
+            y_train = y_train.detach().cpu().numpy()
+
+        # Compute ROC curve
+        fpr, tpr, thresholds = roc_curve(y_train, y_proba)
+
+        # Create a DataFrame from the FPR, TPR, and thresholds
+        roc_df = pd.DataFrame({"FPR": fpr, "TPR": tpr, "Thresholds": thresholds})
+
+        print(roc_df)
+
+        # Define the distance function from each point on the ROC curve to (fpr = 0, tpr = 1)
+        roc_df["Distance"] = np.sqrt((roc_df["FPR"]) ** 2 + (roc_df["TPR"] - 1) ** 2)
+
+        # Find the index of the row with the minimal distance
+        min_distance_index = roc_df["Distance"].idxmin()
+
+        # Get the best threshold at this index
+        best_threshold = roc_df.loc[min_distance_index, "Thresholds"]
+        self.best_threshold = best_threshold
+
+        if self.verbose == 1:
```

Comment:
Can we change this to self.verbose == 2? verbose == 1 was previously used only for textual output